lib/kernel/doc/src/file.xml

```diff
@@ -401,6 +401,52 @@ f.txt:  {person, "kalle", 25}.
         </taglist>
       </desc>
     </func>
+    <func>
+      <name name="del_dir" arity="2" since="OTP 23.0"/>
+      <fsummary>Delete a directory.</fsummary>
+      <desc>
+        <p>Deletes directory <c><anno>Dir</anno></c> as specified by
+          <c><anno>Opts</anno></c>, which contains zero or more
+          of the following options:</p>
+        <taglist>
+          <tag><c>recursive</c></tag>
+          <item>
+            <p>The contents of the directory is recursively deleted
+             before the directory itself is.</p>
+          </item>
+          <tag><c>force</c></tag>
+          <item>
+            <p>Ignores errors when accessing or deleting a file or
+             directory.</p>
+          </item>
+          <tag><c>keeptop</c></tag>
+          <item>
+            <p>The top-level directory is not deleted.</p>
+          </item>
+        </taglist>
+        <p>Returns one of the following:</p>
+        <taglist>
+          <tag><c>ok</c></tag>
+          <item>
+            <p>The operation completed with no non-ignored errors.</p>
+          </item>
+          <tag><c>{error, badarg}</c></tag>
+          <item>
+            <p>An invalid option is present in <anno>Opts</anno>.</p>
+          </item>
+          <tag><c>{error, posix()}</c></tag>
+          <item>
+            <p>An error occurred when accessing or deleting the
+             top-level directory, in non-recursive mode.</p>
+          </item>
+          <tag><c>{error, {posix(), File}}</c></tag>
+          <item>
+            <p>An error occurred when accessing or deleting a file or
+             directory, in recursive mode.</p>
```

Comment:
Should it clarify that, if an error happens in the middle of recursive deletion, the directory will be in a state where part of files were deleted and the other part was not?

Reply:
I would think that's obvious, but if OTP wants that clarification I'll do it.

Reply:
It's obvious that the file that couldn't be deleted will be left alone, but I'm not sure it's as obvious that it exits early on the first error.

If you run `rm -r` on a folder that has a bunch of files that the user has permission to delete and some files they lack permission for, then only the latter will be kept. When using `file:del_dir/2` it might exit early and keep some of the files that could've been deleted if not for the error.

I think both approaches are reasonable, but if we exit early I'd like the documentation to say that it may leave things half-finished.

Reply:
I agree that bailing out early would be against users' expectations.  In that case the reasonable thing would be to imply 'force' in the recursive calls, and just return whatever del_dir/1 returns on the top-level directory; if it's non-empty we get eexist.  We will however not get any indication about which file(s) it failed to remove.  I'm not sure I'd want to collect a list of all failed files and their specific errors.  (The user could list the directory themselves if they really want to know.)

Reply:
> and just return whatever del_dir/1 returns on the top-level directory; if it's non-empty we get eexist. We will however not get any indication about which file(s) it failed to remove

How would this interact with `keeptop`? Would there be files left in the directory without an indication of error?

Reply:
The short answer is that I've decided to remove the `keeptop` option.

I started removing the `recursive` option and implying `force` in recursive deletes, and it became clear that `force` too was entirely redundant, as it only affects whether a failure to delete the top-level component returns `ok` on an `error`-tuple.  And the `keeptop` option would require more code than I though it worth.  So in the end I reduced it to `del_dir_r/1` with no options and equivalent to `rm -rf`.